### ISSUE TITLE
Disconnect hangs forever RECEIPT frame is not received from server

### ISSUE THREAD

**Issue by dselim** [go-stomp/stomp#122]
As STOMP protocol specification mentions:

> Note that, if the server closes its end of the socket too quickly, the client might never receive the expected RECEIPT frame. See the [Connection Lingering](https://stomp.github.io/stomp-specification-1.2.html#Connection_Lingering) section for more information.

I've encountered a few instances of `(*Conn).Disconnect` hanging forever because no `RECEIPT` frame was received from the server. Perhaps some form of timeout should be added to this channel read?

https://github.com/go-stomp/stomp/blob/e752104b359f6ce31f098a630d01999a41f33173/conn.go#L424

**Comment by worg** [comment]
Yes, the `RECEIPT` handling is funky (another error is that it assumes all receipts are disconnects which is not always the case).
I'll add a `ConnOpt` to enable a timeout there

**Comment by miguelsorianod** [comment]
Hi,

Has there been some progress around this? Is there a workaround for it?

I am experiencing a similar issue when `Unsubscribe` is called.
Debugging the code it seems it ends up locked in `s.closeCond.Wait()` of the close method.

I can try to provide more information if needed.

Thanks

**Comment by worg** [comment]
Hey, not much progress sorry, my day job no loner requires STOMP and
unfortunately I haven't been able to make time to dig further into this
issue. nonetheless PRs are very welcome if anyone wants to tackle this

**Comment by jstangroome** [comment]
>I am experiencing a similar issue when Unsubscribe is called.

I hit this exact issue too. In my case my code was calling `Unsubscribe()` on the same Goroutine that was also responsible for consuming incoming messages and by moving the call to `Unsubscribe()` to another Goroutine, this issue appears to have been mitigated.

The issue only occurred where message throughput was high enough that other messages arrived between sending `UNSUBSCRIBE` and receiving `RECEIPT`.

My theory is that the `processLoop` gets stuck writing to a Subscription's unbuffered channel here:
https://github.com/go-stomp/stomp/blob/e752104b359f6ce31f098a630d01999a41f33173/conn.go#L335

By ensuring the app continues to consume all messages coming in on all Subscription channels, the `processLoop` can continue processing the Connection's `readCh` and eventually handle the `RECEIPT` message, close the channel, `Broadcast()` on `closeCond`, and return from `Unsubscribe()`.

https://github.com/go-stomp/stomp/blob/e752104b359f6ce31f098a630d01999a41f33173/conn.go#L305-L310
